### PR TITLE
Add a new buildcheck that encourages MSBuild logic authors to dispose of unused Private Items from their Targets

### PR DESCRIPTION
**Pull request by baronfel**
# Add BC0303 BuildCheck: Detect undisposed private item lists in targets

## Summary

Adds a new BuildCheck analyzer (BC0303) that detects memory waste from undisposed private item lists in MSBuild targets.

## Problem

MSBuild targets often create temporary "private" item lists (conventionally prefixed with `_`) for internal calculations. These items persist in memory for the duration of the build unless explicitly removed, which can waste memory in large builds.

**Example of problematic code:**
```xml
<Target Name="BadExample">
  <ItemGroup>
    <_DotnetExecArgs Include="dotnet;tool;exec;$(CoolToolName)" />
  </ItemGroup>
  <Exec Command="@(_DotnetExecArgs, ' ')" />
  <!-- Missing: <_DotnetExecArgs Remove="@(_DotnetExecArgs)" /> -->
</Target>
```

## Solution

This PR introduces **BC0303 - ItemDisposalCheck**, a new built-in BuildCheck analyzer that:

- Detects private item lists (starting with `_`) created within targets via `Include`
- Verifies they are cleaned up with a corresponding `Remove` operation
- Exempts items exposed via `Outputs` or `Returns` attributes (part of the target's public contract)

### Implementation
- **File**: `src/Build/BuildCheck/Checks/ItemDisposalCheck.cs`
- **Severity**: Warning (default)
- **Scope**: Project file only
- **Key Features**:
  - Order-sensitive analysis (Remove must come after Include)
  - Case-insensitive item type matching
  - Uses `ExpressionShredder.GetReferencedItemNamesAndMetadata()` for robust parsing of Outputs/Returns expressions
  - Handles complex MSBuild expressions including transforms, separators, item functions, and nested properties

### Why ExpressionShredder instead of Regex?
Initially considered regex-based parsing, but switched to MSBuild's built-in `ExpressionShredder` to properly handle:
- Complex transforms with semicolons: `@(_Items->'%(Identity);%(FullPath)')`
- Multiple item references: `@(_Sources);@(_Resources);@(_Content)`
- Custom separators: `@(_Files, ';')`
- Item functions: `@(_Duplicates->Distinct())`
- All other MSBuild expression features

### Documentation
- **Error Code**: BC0303
- **Documentation**: `documentation/specs/BuildCheck/Codes.md`
- **Specification**: `documentation/specs/BuildCheck/ItemDisposalCheck.md`

## Testing

Added 31 comprehensive test cases in `src/BuildCheck.UnitTests/ItemDisposalCheck_Tests.cs`:

**Positive tests** (should fire):
- Private items not removed
- Multiple undisposed items
- Conditional includes without cleanup

**Negative tests** (should NOT fire):
- Properly removed items
- Public items (no `_` prefix)
- Items in `Outputs` or `Returns`
- Update-only operations

**Edge cases**:
- Case-insensitive matching
- Remove before Include (doesn't clean up the Include)
- Complex Outputs/Returns expressions with transforms and semicolons
- Multiple item references in single expression
- Item functions in Returns

All 31 tests pass ✅

## Related Work

This analyzer was created using the msbuild-buildcheck-creator agent, following established patterns in the BuildCheck infrastructure.

### CODE REVIEW

**Comment by AraHaan** [comment]
As long as this results in a compile error to force them to dispose of such wasted memory when finished with it. That way they are forced to fix their code with no way to disable this error.

**Comment by baronfel** [comment]
@AraHaan it definitely will not be an error in the initial rollout - but like all MSBuild diagnostics it will be able to be _treated_ as an error via `TreatWarningsAsError`, and via the buildcheck-specific editorconfig support.

File: .claude/agents/msbuild-buildcheck-creator.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
+---
+name: msbuild-buildcheck-creator
```
Comment (YuliiaKovalova) [inline .claude/agents/msbuild-buildcheck-creator.md:2]: is this instruction file needed? or should it be more generic (not claude)?

File: src/Build/BuildCheck/Checks/ItemDisposalCheck.cs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        {
+            AnalyzeTarget(target, context);
+        }
+    }
+
+    /// <summary>
+    /// Analyzes a single target for private items that are not properly disposed.
+    /// </summary>
+#pragma warning disable CS0618 // Type or member is obsolete - ParsedItemsCheckData is obsolete but we need it for the hook
+    private void AnalyzeTarget(ProjectTargetElement target, BuildCheckDataContext<ParsedItemsCheckData> context)
```
Comment (YuliiaKovalova) [inline src/Build/BuildCheck/Checks/ItemDisposalCheck.cs]: can EvaluatedItemsData be used instead of ParsedItemsCheckData?

Comment (YuliiaKovalova) [reply src/Build/BuildCheck/Checks/ItemDisposalCheck.cs]: i fixed it :)
